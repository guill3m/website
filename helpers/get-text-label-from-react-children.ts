import { Children, type ReactNode } from 'react'

export default function getTextLabelFromReactChildren(
  children: ReactNode | ReactNode[],
): string {
  let label = ''

  Children.map(children, (child) => {
    if (typeof child === 'string') {
      label += child
    }
  })

  return label.trim()
}
