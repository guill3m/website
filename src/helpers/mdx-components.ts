import Link from '../components/link.astro'

export const components = { a: Link }
