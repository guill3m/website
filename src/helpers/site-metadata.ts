const siteMetadata = {
	author: {
		name: 'Guillem Andreu',
		social: {
			bluesky: 'guill3m.me',
			github: 'guill3m',
			lastfm: 'guill3m',
			linkedin: 'guillemandreu',
		},
	},
	host: 'www.guillemandreu.com',
}

export default siteMetadata
