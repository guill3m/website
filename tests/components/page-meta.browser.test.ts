import { describe, expect, test } from 'vitest'
import { renderAstroComponent } from '../helpers'

import PageMetadata from '../../src/components/page-meta.astro'

describe('Page Metadata component', () => {
	test('minimal props', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		expect(result.querySelector('title')).toHaveTextContent(
			/^Test page - Guillem Andreu$/,
		)
		expect(result.querySelector('meta[property=og:title]')).toHaveAttribute(
			'content',
			'Test page - Guillem Andreu',
		)
		expect(result.querySelector('link[rel=canonical]')).toHaveAttribute(
			'href',
			'https://test.dev/example/',
		)
		expect(result.querySelector('meta[property=og:type]')).toHaveAttribute(
			'content',
			'website',
		)
	})

	test('with non-localised alternate', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					alternates: [
						{
							href: '/example-alt/',
						},
					],
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		const alternates = result.querySelectorAll('link[rel=alternate]')
		expect(alternates).toHaveLength(1)
		expect(alternates[0]).toHaveAttribute(
			'href',
			'https://test.dev/example-alt/',
		)
		expect(alternates[0]).not.toHaveAttribute('hreflang')
	})

	test('with external non-localised alternate', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					alternates: [
						{
							href: 'https://test-alt.dev/example/',
						},
					],
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		const alternates = result.querySelectorAll('link[rel=alternate]')
		expect(alternates).toHaveLength(1)
		expect(alternates[0]).toHaveAttribute(
			'href',
			'https://test-alt.dev/example/',
		)
		expect(alternates[0]).not.toHaveAttribute('hreflang')
	})

	test('with external non-localised alternate as URL object', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					alternates: [
						{
							href: new URL('/example/', 'https://test-alt.dev'),
						},
					],
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		const alternates = result.querySelectorAll('link[rel=alternate]')
		expect(alternates).toHaveLength(1)
		expect(alternates[0]).toHaveAttribute(
			'href',
			'https://test-alt.dev/example/',
		)
		expect(alternates[0]).not.toHaveAttribute('hreflang')
	})

	test('with localised alternate', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					alternates: [
						{
							href: '/de/example/',
							lang: 'de',
						},
					],
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		const alternates = result.querySelectorAll('link[rel=alternate]')
		expect(alternates).toHaveLength(3)
		expect(alternates[0]).toHaveAttribute('href', 'https://test.dev/example/')
		expect(alternates[0]).toHaveAttribute('hreflang', 'x-default')
		expect(alternates[1]).toHaveAttribute('href', 'https://test.dev/example/')
		expect(alternates[1]).toHaveAttribute('hreflang', 'en')
		expect(alternates[2]).toHaveAttribute(
			'href',
			'https://test.dev/de/example/',
		)
		expect(alternates[2]).toHaveAttribute('hreflang', 'de')
	})

	test('with default localised alternate', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					alternates: [
						{
							isDefault: true,
							href: '/example/',
							lang: 'en',
						},
						{
							href: '/es/example/',
							lang: 'es',
						},
					],
				},
				lang: 'de',
			},
			request: new Request(`https://test.dev/de/example/`),
		})

		expect(result).toMatchSnapshot()

		const alternates = result.querySelectorAll('link[rel=alternate]')
		expect(alternates).toHaveLength(4)
		expect(alternates[0]).toHaveAttribute('href', 'https://test.dev/example/')
		expect(alternates[0]).toHaveAttribute('hreflang', 'x-default')
		expect(alternates[1]).toHaveAttribute(
			'href',
			'https://test.dev/de/example/',
		)
		expect(alternates[1]).toHaveAttribute('hreflang', 'de')
		expect(alternates[2]).toHaveAttribute('href', 'https://test.dev/example/')
		expect(alternates[2]).toHaveAttribute('hreflang', 'en')
		expect(alternates[3]).toHaveAttribute(
			'href',
			'https://test.dev/es/example/',
		)
		expect(alternates[3]).toHaveAttribute('hreflang', 'es')
	})

	test('with overwritten title and description', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					title: 'Meta title',
					description: 'Meta description',
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		expect(result.querySelector('title')).toHaveTextContent(/^Meta title$/)
		expect(result.querySelector('meta[property=og:title]')).toHaveAttribute(
			'content',
			'Meta title',
		)
		expect(result.querySelector('meta[name=description]')).toHaveAttribute(
			'content',
			'Meta description',
		)
		expect(
			result.querySelector('meta[property=og:description]'),
		).toHaveAttribute('content', 'Meta description')
	})

	test('with OpenGraph properties', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					openGraph: {
						type: 'article',
						title: 'OpenGraph title',
						description: 'OpenGraph description',
					},
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		expect(result.querySelector('meta[property=og:type]')).toHaveAttribute(
			'content',
			'article',
		)
		expect(result.querySelector('title')).toHaveTextContent(
			/^Test page - Guillem Andreu$/,
		)
		expect(result.querySelector('meta[property=og:title]')).toHaveAttribute(
			'content',
			'OpenGraph title',
		)
		expect(
			result.querySelector('meta[property=og:description]'),
		).toHaveAttribute('content', 'OpenGraph description')
	})

	test('with overwritten title and description and OpenGraph properties', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					title: 'Meta title',
					description: 'Meta description',
					openGraph: {
						type: 'article',
						title: 'OpenGraph title',
						description: 'OpenGraph description',
					},
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		expect(result.querySelector('meta[property=og:type]')).toHaveAttribute(
			'content',
			'article',
		)
		expect(result.querySelector('title')).toHaveTextContent(/^Meta title$/)
		expect(result.querySelector('meta[property=og:title]')).toHaveAttribute(
			'content',
			'OpenGraph title',
		)
		expect(result.querySelector('meta[name=description]')).toHaveAttribute(
			'content',
			'Meta description',
		)
		expect(
			result.querySelector('meta[property=og:description]'),
		).toHaveAttribute('content', 'OpenGraph description')
	})

	test('with image', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					image: {
						alt: 'Alt text',
						src: '/assets/page-thumbnail.png',
						height: 300,
						width: 900,
					},
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()

		expect(result.querySelector('meta[property=og:image]')).toHaveAttribute(
			'content',
			'https://test.dev/assets/page-thumbnail.png',
		)
		expect(result.querySelector('meta[property=og:image:alt]')).toHaveAttribute(
			'content',
			'Alt text',
		)
		expect(
			result.querySelector('meta[property=og:image:height]'),
		).toHaveAttribute('content', '300')
		expect(
			result.querySelector('meta[property=og:image:width]'),
		).toHaveAttribute('content', '900')
	})

	test('all props', async () => {
		const result = await renderAstroComponent(PageMetadata, {
			props: {
				title: 'Test page',
				meta: {
					canonical: new URL('https://production.dev/example/'),
					title: 'Meta title',
					description: 'Meta description',
					alternates: [
						{
							href: '/example-alt/',
						},
					],
					image: {
						alt: 'Alt text',
						src: '/assets/page-thumbnail.png',
						height: 300,
						width: 900,
					},
					openGraph: {
						type: 'article',
						title: 'OpenGraph title',
						description: 'OpenGraph description',
					},
				},
			},
			request: new Request(`https://test.dev/example/`),
		})

		expect(result).toMatchSnapshot()
	})
})
