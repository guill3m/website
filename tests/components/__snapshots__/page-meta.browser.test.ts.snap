// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page Metadata component > all props 1`] = `
<DocumentFragment>
  <title>
    Meta title
  </title>
  <meta
    content="OpenGraph title"
    property="og:title"
  />
  <meta
    content="Meta description"
    name="description"
  />
  <meta
    content="OpenGraph description"
    property="og:description"
  />
  <link
    href="https://production.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://production.dev/example/"
    property="og:url"
  />
  <link
    href="https://test.dev/example-alt/"
    rel="alternate"
  />
  <meta
    content="article"
    property="og:type"
  />
  <meta
    content="https://test.dev/assets/page-thumbnail.png"
    property="og:image"
  />
  <meta
    content="Alt text"
    property="og:image:alt"
  />
  <meta
    content="300"
    property="og:image:height"
  />
  <meta
    content="900"
    property="og:image:width"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > minimal props 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with OpenGraph properties 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="OpenGraph title"
    property="og:title"
  />
  <meta
    content="OpenGraph description"
    property="og:description"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <meta
    content="article"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with default localised alternate 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/de/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/de/example/"
    property="og:url"
  />
  <link
    href="https://test.dev/example/"
    hreflang="x-default"
    rel="alternate"
  />
  <link
    href="https://test.dev/de/example/"
    hreflang="de"
    rel="alternate"
  />
  <link
    href="https://test.dev/example/"
    hreflang="en"
    rel="alternate"
  />
  <link
    href="https://test.dev/es/example/"
    hreflang="es"
    rel="alternate"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with external non-localised alternate 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <link
    href="https://test-alt.dev/example/"
    rel="alternate"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with external non-localised alternate as URL object 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <link
    href="https://test-alt.dev/example/"
    rel="alternate"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with image 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <meta
    content="website"
    property="og:type"
  />
  <meta
    content="https://test.dev/assets/page-thumbnail.png"
    property="og:image"
  />
  <meta
    content="Alt text"
    property="og:image:alt"
  />
  <meta
    content="300"
    property="og:image:height"
  />
  <meta
    content="900"
    property="og:image:width"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with localised alternate 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <link
    href="https://test.dev/example/"
    hreflang="x-default"
    rel="alternate"
  />
  <link
    href="https://test.dev/example/"
    hreflang="en"
    rel="alternate"
  />
  <link
    href="https://test.dev/de/example/"
    hreflang="de"
    rel="alternate"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with non-localised alternate 1`] = `
<DocumentFragment>
  <title>
    Test page - Guillem Andreu
  </title>
  <meta
    content="Test page - Guillem Andreu"
    property="og:title"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <link
    href="https://test.dev/example-alt/"
    rel="alternate"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with overwritten title and description 1`] = `
<DocumentFragment>
  <title>
    Meta title
  </title>
  <meta
    content="Meta title"
    property="og:title"
  />
  <meta
    content="Meta description"
    name="description"
  />
  <meta
    content="Meta description"
    property="og:description"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <meta
    content="website"
    property="og:type"
  />
</DocumentFragment>
`;

exports[`Page Metadata component > with overwritten title and description and OpenGraph properties 1`] = `
<DocumentFragment>
  <title>
    Meta title
  </title>
  <meta
    content="OpenGraph title"
    property="og:title"
  />
  <meta
    content="Meta description"
    name="description"
  />
  <meta
    content="OpenGraph description"
    property="og:description"
  />
  <link
    href="https://test.dev/example/"
    rel="canonical"
  />
  <meta
    content="https://test.dev/example/"
    property="og:url"
  />
  <meta
    content="article"
    property="og:type"
  />
</DocumentFragment>
`;
